Source/WebCore/html/Origin.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "Origin.h"
+
+#include "DOMURL.h"
+#include "DOMWindow.h"
+#include "ExtendableMessageEvent.h"
+#include "HTMLAnchorElement.h"
+#include "HTMLAreaElement.h"
+#include "JSDOMBindingSecurity.h"
+#include "JSDOMBindingSecurityInlines.h"
+#include "JSDOMGlobalObject.h"
+#include "JSDOMGlobalObjectInlines.h"
+#include "JSDOMURL.h"
+#include "JSDOMWindow.h"
+#include "JSExtendableMessageEvent.h"
+#include "JSHTMLAnchorElement.h"
+#include "JSHTMLAreaElement.h"
+#include "JSMessageEvent.h"
+#include "JSOrigin.h"
+#include "JSWorkerGlobalScope.h"
+#include "LocalDOMWindow.h"
+#include "MessageEvent.h"
+#include "SecurityOrigin.h"
+#include "WorkerGlobalScope.h"
+#include <wtf/URL.h>
+
+namespace WebCore {
+
+inline Origin::Origin(Ref<SecurityOrigin>&& securityOrigin)
+    : m_origin(WTFMove(securityOrigin))
+{
+}
+
+Ref<Origin> Origin::create()
+{
+    return adoptRef(*new Origin(SecurityOrigin::createOpaque()));
+}
+
+Ref<Origin> Origin::create(Ref<SecurityOrigin>&& securityOrigin)
+{
+    return adoptRef(*new Origin(WTFMove(securityOrigin)));
+}
+
+ExceptionOr<Ref<Origin>> Origin::from(ScriptExecutionContext& context, JSC::JSValue value)
+{
+    if (value.isString()) {
+        auto parsedURL = URL { value.getString(context.globalObject()) };
+        if (parsedURL.isValid())
+            return create(SecurityOrigin::create(parsedURL));
```

Comment:
I would tend to write it as:
```
if (value.isString()) {
    if (!parsedURL.isValid())
        return Exception { ExceptionCode::TypeError, "URL is invalid"_s };
   return create(...);
}
if (auto...)
```
This is closer to the coding style of returning early and adding descriptions to exceptions is a good practice.